drivers/auxdisplay/auxdisplay_serlcd.c

```diff
@@ -0,0 +1,419 @@
+/*
+ * Copyright (c) 2023 Jan Henke <Jan.Henke@taujhe.de>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT sparkfun_serlcd
+
+#include <stdlib.h>
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/auxdisplay.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(auxdisplay_serlcd, CONFIG_AUXDISPLAY_LOG_LEVEL);
+
+/*
+ * | in ASCII, used to begin a display command
+ */
+#define SERLCD_BEGIN_COMMAND 0x7C
+
+/*
+ * special command for the underlying display controller
+ */
+#define SERLCD_BEGIN_SPECIAL_COMMAND 0xFE
+
+/*
+ * delay in milliseconds after a normal command was sent
+ */
+#define SERLCD_COMMAND_DELAY 10
+
+/*
+ * delay in milliseconds after a special command was sent
+ */
+#define SERLCD_SPECIAL_COMMAND_DELAY 50
+
+/*
+ * maximum amount of custom chars the display supports
+ */
+#define SERLCD_CUSTOM_CHAR_MAX_COUNT 8
+
+/*
+ * height of a custom char in bits
+ */
+#define SERLCD_CUSTOM_CHAR_HEIGHT 8
+
+/*
+ * width of s custom char in bits
+ */
+#define SERLCD_CUSTOM_CHAR_WIDTH 5
+
+/*
+ * char code for the first customer char
+ */
+#define SERLCD_CUSTOM_CHAR_INDEX_BASE 0x08
+
+/*
+ * bitmask
+ */
+#define SERLCD_CUSTOM_CHAR_BITMASK 0xf8
+
+struct auxdisplay_serlcd_data {
+	bool power;
+	bool cursor;
+	bool blinking;
+	uint16_t cursor_x;
+	uint16_t cursor_y;
+};
+
+struct auxdisplay_serlcd_config {
+	struct auxdisplay_capabilities capabilities;
+	struct i2c_dt_spec bus;
+};
+
+enum auxdisplay_serlcd_command {
+	SERLCD_COMMAND_SET_CUSTOM_CHAR = 0x1B,
+	SERLCD_COMMAND_WRITE_CUSTOM_CHAR = 0x23,
+	SERLCD_COMMAND_CLEAR = 0x2D,
+};
+
+enum auxdisplay_serlcd_special_command {
+	SERLCD_SPECIAL_RETURN_HOME = 0x02,
+	SERLCD_SPECIAL_DISPLAY_CONTROL = 0x08,
+	SERLCD_SPECIAL_SET_DD_RAM_ADDRESS = 0x80,
+};
+
+static int auxdisplay_serlcd_send_command(const struct device *dev,
+					  const enum auxdisplay_serlcd_command command)
+{
+	const struct auxdisplay_serlcd_config *config = dev->config;
+	const uint8_t buffer[2] = {SERLCD_BEGIN_COMMAND, command};
+
+	int rc = i2c_write_dt(&config->bus, buffer, sizeof(buffer));
+
+	k_sleep(K_MSEC(SERLCD_COMMAND_DELAY));
+	return rc;
+}
+
+static int
+auxdisplay_serlcd_send_special_command(const struct device *dev,
+				       const enum auxdisplay_serlcd_special_command command)
+{
+	const struct auxdisplay_serlcd_config *config = dev->config;
+	const uint8_t buffer[2] = {SERLCD_BEGIN_SPECIAL_COMMAND, command};
+
+	int rc = i2c_write_dt(&config->bus, buffer, sizeof(buffer));
+
+	k_sleep(K_MSEC(SERLCD_SPECIAL_COMMAND_DELAY));
+	return rc;
+}
+
+static int auxdisplay_serlcd_send_display_state(const struct device *dev,
+						const struct auxdisplay_serlcd_data *data)
+{
+	enum auxdisplay_serlcd_special_command command = SERLCD_SPECIAL_DISPLAY_CONTROL;
```

Comment:
I think it's a better idea to use a `uint8_t` here, the enum may change size depening on the compiler settings and you certainly don't want that to happen.

Reply:
Done